### ISSUE TITLE
"dataclass" field converter functions are ignored

### ISSUE THREAD

**Issue by Kache** [microsoft/pyright#8403]
**Describe the bug**

The "dataclass" library https://www.attrs.org/en/stable/ allows defining field converter functions. Typechecking of the constructors behave as if there was no converter function.

Ideally, the typechecker would infer the input and output types of the converter function. Then, constructor arg would have the same signature as the converter function input, while producing a field value with same type as the converter function output.

Otherwise, typechecker could ignore fields that are converted, e.g. consider constructor arg accept Any, relaxing the constraint when it can't make the correct determination.

I realize this might be considered an enhancement/feature and/or a library-specific typeshed issue instead of a bug.

**Code or Screenshots**

```py
import datetime as dt

import dateutil.parser  # 2.8.2
from attrs import define, field  # 23.1.0


def conv(val: int) -> str:
    return str(val)


@define(frozen=True, auto_attribs=True)
class Foo:
    x: int
    y: str = field(converter=conv)


@define(frozen=True, auto_attribs=True)
class Bar:
    x: int
    y: dt.datetime = field(converter=dateutil.parser.parse)


def build(y: int, s: str):
    return (
        Foo(1, y),  # typecheck error here
        Bar(1, s),  # typecheck error here
    )


print(build(2, '2020-01-01'))
```

Results in:

```
/test_attrs.py
  /test_attrs.py:25:16 - error: Argument of type "int" cannot be assigned to parameter "y" of type "str" in function "__init__"
    "int" is incompatible with "str" (reportArgumentType)
  /test_attrs.py:26:16 - error: Argument of type "str" cannot be assigned to parameter "y" of type "datetime" in function "__init__"
    "str" is incompatible with "datetime" (reportArgumentType)
2 errors, 0 warnings, 0 informations
```

**VS Code extension or command-line**

 * Noticed this when recently upgrading to pyright version 1.1.371 from old version 1.1.350
 * Maybe related to #7713?

**Comment by erictraut** [comment]
The typing spec currently does not recognize `converter` for [dataclass_transform](https://typing.readthedocs.io/en/latest/spec/dataclasses.html#the-dataclass-transform-decorator), so this isn't a bug.

The good news is that there's a [draft modification](https://discuss.python.org/t/discuss-adding-converter-to-dataclass-transform/55125/4) that proposes to add this functionality to the typing spec. This proposal hasn't yet been approved, but it has been submitted to the typing council for final approval. Two of the five typing council members (myself included) have already [signed off](https://github.com/python/typing-council/issues/33), and I expect that we'll hear from the other three members in the near future.

Pyright already has experimental support for this proposal. You can enable it by setting `enableExperimentalFeatures` to true. If and when the proposal is approved and incorporated into the typing spec, I'll move it out of experimental status. If the proposal is rejected, I will remove the experimental support.

**Comment by hynek** [comment]
Could it be that this used to work but has then been gated in 1.1.359? Because this broke the attrs CI and I can't find anything about it in the changelog and it was a tad confusing. I thought for a moment y'all removed it because PEP 712 has been rejected.

**Comment by erictraut** [comment]
Yes, prior to 1.1.359, it was not marked experimental. I moved it to an experimental feature because I didn't want anyone to take a dependency on it unknowingly — in case we needed to remove it.
